### PR TITLE
Implement Grade Enchant User Interface

### PR DESCRIPTION
**Pull request by Asheraf**
### Pull Request Prelude

- [x] I have followed [proper Hercules code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR may be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed

This PR adds support for Grade Enchanter.
image

The missing parts:

- Values in the database are not official and I just guessed them.
- Item protection for older clients is not supported

**Issues addressed:** 

[cont]: https://github.com/HerculesWS/Hercules/blob/master/CONTRIBUTING.md
[code]: https://github.com/HerculesWS/Hercules/wiki/Coding-Style

### CODE REVIEW

File: src/map/clif.c [hunk 1]
```diff
@@ -25151,6 +25158,123 @@ static void clif_sayDialogAlign(struct map_session_data *sd, int npcid, enum say
...
+{
+#if PACKETVER_MAIN_NUM >= 20191016 || PACKETVER_RE_NUM >= 20191016 || PACKETVER_ZERO_NUM >= 20191008
+	sd->state.grade_ui = 0;
+#endif
+}
+
+static void clif_grade_enchant_add_item_result_success(struct map_session_data *sd, int idx, const struct s_grade_info *gi)
+{
+#if PACKETVER_MAIN_NUM >= 20200916 || PACKETVER_RE_NUM >= 20200723
+	nullpo_retv(sd);
```
Comment (4144) [inline src/map/clif.c:25172]: missing nullpo check for ``gi``

File: src/map/grader.c [hunk 2]
```diff
@@ -0,0 +1,406 @@
...
+			++count;
+	}
+
+	return (count > 0);
+}
+
+static bool grader_read_db_libconfig_sub_material(const struct config_setting_t *it, struct grade_material *gm, int n, const char *source)
+{
+	nullpo_ret(it);
+	nullpo_ret(source);
```
Comment (4144) [inline src/map/grader.c:133]: missing check for nullpo for ``gm``

**Review by 4144 — CHANGES_REQUESTED** [review]

**Review by MishimaHaruna — CHANGES_REQUESTED** [review]

File: src/map/log.c [hunk 3]
```diff
@@ -76,6 +76,7 @@ static char log_picktype2char(e_log_pick_type type)
 		case LOG_TYPE_REFINE:           return '2';  // Refine
 		case LOG_TYPE_OTHER:            return 'X';  // Other
 		case LOG_TYPE_ACHIEVEMENT:      return '3';  // Achievement
+		case LOG_TYPE_GRADE:            return '4';  // Grade Enchant
```
[the remark this answers is not in the document]
Comment (Asheraf) [reply src/map/log.c:79]: Added to picklog and provided an update sql file.

**Review by MishimaHaruna — APPROVED** [review]
I also added the type '4' to the zenylog table since there's a `pc->payzeny()` call with the `LOG_TYPE_GRADE` type

**Pull request merged**